### PR TITLE
engine(servermaster): remove embedded etcd

### PR DESCRIPTION
**Pull request by gozssky**
### What problem does this PR solve?

Issue Number: close https://github.com/pingcap/tiflow/issues/6388

### What is changed and how it works?

This PR removes the embedded etcd from servermaster. After this modification, severmaster becomes stateless and it is easy to deploy and maintain. As a workaround, servermaster need an external etcd to support leader election and service discovery. In the future, I will implement leader election and service discovery on metastore which using MySQL.

#### Tests 

 - Unit test
 - Integration test
 - Manual test (add detailed scripts or steps below)
 - No code

### Release note 

```release-note
None
```

### CODE REVIEW

File: engine/deployments/docker-compose/1m1e.yaml [hunk 1]
```diff
@@ -8,27 +8,19 @@ services:
...
-                        - "server-master:10240"
-                        - "--peer-urls"
-                        - "http://127.0.0.1:8291"
-                        - "--advertise-peer-urls"
-                        - "http://server-master:8291"
-                        - "--frame-meta-endpoints"
-                        - "frame-mysql-standalone:3306"
-                        - "--business-meta-endpoints"
-                        - "business-etcd-standalone:2379"
+                        - "--master-addr=0.0.0.0:10240"
```
Comment (lonng) [inline engine/deployments/docker-compose/1m1e.yaml:13]: Ref https://github.com/pingcap/tiflow/pull/6455. The command line parameters should be changed.

File: engine/deployments/docker-compose/1m1e.yaml [hunk 2]
```diff
@@ -38,15 +30,11 @@ services:
...
+                        - "--config=/config.toml"
                         - "--log-level=debug"
-                        - "--join"
-                        - "server-master:10240"
-                        - "--worker-addr"
-                        - "0.0.0.0:10241"
-                        - "--advertise-addr"
-                        - "server-executor:10241"
+                        - "--join=server-master:10240"
+                        - "--worker-addr=0.0.0.0:10241"
```
Comment (lonng) [inline engine/deployments/docker-compose/1m1e.yaml:36]: ditto

File: engine/deployments/docker-compose/1m1e.yaml (hunk 1 above)
Comment (gozssky) [reply engine/deployments/docker-compose/1m1e.yaml]: Done.

**Comment by gozssky** [comment]
Chaos test passed: https://github.com/gozssky/tiflow/actions/runs/2743071392.

**Comment by amyangfei** [comment]
> Chaos test passed: https://github.com/gozssky/tiflow/actions/runs/2743071392.

I checked server master log and git hash, it seems running with master branch

```go
[2022/07/26 23:15:02.457 +00:00] [INFO] [version.go:35] ["Welcome to Dataflow Engine"] [release-version=v6.2.0-master] [git-hash=7c9c143c4f07e41c6a0c0360d8976b1c2251f5dc] [git-branch=master] [utc-build-time="2022-07-26 23:09:04"] [go-version="go version go1.18.4 linux/amd64"]
```

File: engine/servermaster/campaign.go [hunk 3]
```diff
@@ -192,19 +177,13 @@ func (s *Server) checkLeaderExists(ctx context.Context) (retry bool, err error)
 		}
 
 		// step-3, case-c
```
Comment (amyangfei) [inline engine/servermaster/campaign.go:165]:
```suggestion
		// step-2, case-c
```

File: engine/servermaster/campaign.go [hunk 4]
```diff
@@ -192,19 +177,13 @@ func (s *Server) checkLeaderExists(ctx context.Context) (retry bool, err error)
...
-		leader.IsEtcdLeader = true
+		leader.IsLeader = true
 		log.Info("start to watch server master leader",
 			zap.String("leader-name", leader.Name), zap.String("addr", leader.AdvertiseAddr))
-		s.watchLeader(ctx, leader, rev)
+		s.watchLeader(ctx, leader, key, rev)
 		log.Info("server master leader changed")
 	}
 
-	// step-3
```
Comment (amyangfei) [inline engine/servermaster/campaign.go:203]: Can remove `step-3` in the comment of function `checkLeaderExists`

**Comment by gozssky** [comment]
> > Chaos test passed: https://github.com/gozssky/tiflow/actions/runs/2743071392.
> 
> I checked server master log and git hash, it seems running with master branch
> 
> ```go
> [2022/07/26 23:15:02.457 +00:00] [INFO] [version.go:35] ["Welcome to Dataflow Engine"] [release-version=v6.2.0-master] [git-hash=7c9c143c4f07e41c6a0c0360d8976b1c2251f5dc] [git-branch=master] [utc-build-time="2022-07-26 23:09:04"] [go-version="go version go1.18.4 linux/amd64"]
> ```

@amyangfei Sorry, the correct link is https://github.com/gozssky/tiflow/actions/runs/2739427332.

File: engine/servermaster/campaign.go [hunk 5]
```diff
@@ -104,22 +100,11 @@ func (s *Server) leaderLoop(ctx context.Context) error {
 		leaderResignFn = newResignFn
 
 		errg, leaderCtx := errgroup.WithContext(newCtx)
-		// Note there isn't any correctness issue if server master leader is
-		// different from the etcd leader. But in order to reduce the maintenance
-		// cost, we evict server master leader actively if etcd leader changes.
 		errg.Go(func() error {
```
[the remark this answers is not in the document]
Comment (gozssky) [reply engine/servermaster/campaign.go]: Done.

File: engine/servermaster/server.go [hunk 6]
```diff
@@ -227,13 +226,11 @@ func (s *Server) Heartbeat(ctx context.Context, req *pb.HeartbeatRequest) (*pb.H
 
 	resp, err := s.executorManager.HandleHeartbeat(req)
 	if err == nil && resp.Err == nil {
-		s.members.RLock()
-		defer s.members.RUnlock()
-		addrs := make([]string, 0, len(s.members.m))
-		for _, member := range s.members.m {
-			addrs = append(addrs, member.AdvertiseAddr)
+		for _, nodeInfo := range s.discoveryKeeper.Snapshot() {
```
Comment (amyangfei) [inline engine/servermaster/server.go:229]: Could `s.discoveryKeeper.Snapshot()` returns `nil` here, which will cause the client list of master client update to empty.

Comment (gozssky) [reply engine/servermaster/server.go:229]: it's unlikely that it returns nil here. For safety, we add an extra check here.

File: engine/servermaster/campaign.go (hunk 4 above)
Comment (gozssky) [reply engine/servermaster/campaign.go:203]: Done.

**Review by amyangfei — APPROVED** [review]

File: engine/chaos/cases/case_fake_job.go [hunk 7]
```diff
@@ -20,9 +20,9 @@ import (
 	"fmt"
 	"time"
 
+	"github.com/pingcap/log"
```
Comment (lonng) [inline engine/chaos/cases/case_fake_job.go:23]: Why move this line from the third group here?

Comment (gozssky) [reply engine/chaos/cases/case_fake_job.go:23]: For tiflow, this line is the third-party package.

**Review by maxshuang — APPROVED** [review]

**Pull request merged**